pkg/lib/saml/service.go

```diff
@@ -892,6 +887,63 @@ func (s Service) recordSessionParticipant(
 	return nil
 }
 
+func (s *Service) ResolveUserAttributes(sp *config.SAMLServiceProviderConfig, userInfo map[string]interface{}) ([]samlprotocol.Attribute, error) {
+	attrs := []samlprotocol.Attribute{
+		{
+			FriendlyName: "User ID",
+			Name:         "sub",
+			NameFormat:   samlprotocol.SAMLAttrnameFormatBasic,
+			Values: []samlprotocol.AttributeValue{{
+				Type:  samlprotocol.SAMLAttrTypeString,
+				Value: userInfo["sub"].(string),
+			}},
+		},
+	}
+
+	valuesMap := map[string][]string{}
+	for _, mapping := range sp.Attributes.Mappings {
+		values := []string{}
+		switch {
+		case mapping.From.UserProfile != nil:
+			jsonPointer := mapping.From.UserProfile.MustGetJSONPointer()
+			raw, err := jsonPointer.Traverse(userInfo)
+			if err != nil {
+				// If the attribute does not exist, just skip
+			} else {
+				values, err = formatAttribute(raw)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+		valuesMap[mapping.To.SAMLAttribute] = values
+	}
+
+	for _, attrDef := range sp.Attributes.Definitions {
+		values, ok := valuesMap[attrDef.Name]
+		if !ok {
+			continue
+		}
+		attrValues := []samlprotocol.AttributeValue{}
+		for _, val := range values {
+			attrValues = append(attrValues,
+				samlprotocol.AttributeValue{
+					Type:  samlprotocol.SAMLAttrTypeString,
```

Comment:
I think we should support primitive types and output the correct type. For example, if the resolved user profile attribute is a string, then the XML type is string.

We need to reference https://www.w3.org/TR/xmlschema-2/#built-in-primitive-datatypes and specify in the spec that what Go data type map to which XML data type.

Reply:
I've modified the spec in https://github.com/authgear/authgear-server/pull/5072/commits/c334fe5549a91f82f2c9cfea486fce8479554109

Documented the mappings between JSON data type and SAML attribute type when using simple mapping.